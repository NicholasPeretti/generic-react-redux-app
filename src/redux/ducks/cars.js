import { Map } from 'immutable'

export const NAMESPACE = 'CARS'
export const SET_FETCHING = `${NAMESPACE} SET_FETCHING`
export const SET_PAGE = `${NAMESPACE} SET_PAGE`
export const SET_CARS = `${NAMESPACE} SET_CARS`
export const SET_RESULTS_COUNT = `${NAMESPACE} SET_RESULTS_COUNT`
export const SET_ERROR = `${NAMESPACE} SET_ERROR`

export const setFetching = (fetching) => ({
  type: SET_FETCHING,
  fetching: !!fetching
})

export const setPage = (page = 1) => ({
  type: SET_PAGE,
  page
})

export const setCars = (cars = []) => ({
  type: SET_CARS,
  cars
})

export const setResultsCount = (resultsCount = 0) => ({
  type: SET_RESULTS_COUNT,
  resultsCount
})

export const setError = (error) => ({
  type: SET_ERROR,
  error
})

export const getState = (state) => (
  state[NAMESPACE]
)

export const getCars = (state) => (
  getState(state).cars
)

export const getPage = (state) => (
  getState(state).page
)

export const isFetching = (state) => (
  getState(state).fetching
)

export const defaultState = {
  cars: [],
  carsMap: new Map(),
  fetching: false,
  page: 1,
  resultsCount: 0,
  error: null
}

export default function reducer (state = defaultState, action = {}) {
  switch (action.type) {
    case SET_FETCHING: {
      state = Object.assign({}, state, {
        fetching: action.fetching
      })
      break
    }

    case SET_PAGE: {
      state = Object.assign({}, state, {
        page: action.page
      })
      break
    }

    case SET_CARS: {
      const { cars } = action
      var carsMap = new Map()
      cars.forEach(car => {
        carsMap = carsMap.set(car.stockNumber, car)
      })

      state = Object.assign({}, state, {
        cars,
        carsMap
      })
      break
    }

    case SET_RESULTS_COUNT: {
      state = Object.assign({}, state, {
        resultsCount: action.resultsCount
      })
      break
    }

    case SET_ERROR: {
      state = Object.assign({}, state, {
        error: action.error
      })
      break
    }
  }

  return state
}
