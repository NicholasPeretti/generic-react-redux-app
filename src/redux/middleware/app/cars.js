import { isApiAction, API_REQUEST, API_SUCCESS, API_ERROR, apiRequest } from '../core/api'
import { NAMESPACE, SET_MILEAGE_SORT, setFetching, setCars, setResultsCount, setError, setPage, getMileageSort } from '../../ducks/cars'
import { GET_CARS } from '../../endpoints'

export const fetchPage = (page = 1) => {
  const { url, method } = GET_CARS
  return apiRequest(url, method, NAMESPACE, { page })
}

export default function middleware (store) {
  return dispatch => action => {
    if (action.type === SET_MILEAGE_SORT) {
      dispatch(action)
      store.dispatch(fetchPage(1))
      return
    }

    if (!isApiAction(action, NAMESPACE)) {
      dispatch(action)
      return
    }

    switch (action.type) {
      case API_REQUEST: {
        let state = store.getState()
        let mileageSort = getMileageSort(state)

        if (mileageSort) {
          action.payload.sort = mileageSort > 0 ? 'asc' : 'desc'
        }

        dispatch(action)
        dispatch(setFetching(true))
        dispatch(setPage(action.payload.page))
        break
      }

      case API_SUCCESS: {
        dispatch(setFetching(false))
        dispatch(setCars(action.payload.cars))
        dispatch(setResultsCount(action.payload.totalPageCount))
        break
      }

      case API_ERROR: {
        dispatch(setFetching(false))
        dispatch(setError(action.payload))
        break
      }
    }
  }
}
