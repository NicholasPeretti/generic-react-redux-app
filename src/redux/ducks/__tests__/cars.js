import DuckTest from '../../../utils/tests/duck'
import { Map } from 'immutable'
import * as Cars from '../cars'

const MOCK_CARS = [
  {
    stockNumber: 10056,
    manufacturerName: 'Skoda',
    modelName: 'Roomster',
    color: 'white',
    mileage: {
      number: 186087,
      unit: 'km'
    },
    fuelType: 'Petrol',
    pictureUrl: 'http://localhost:3001/car.svg'
  },
  {
    stockNumber: 10205,
    manufacturerName: 'Mercedes-Benz',
    modelName: 'SLR McLaren',
    color: 'red',
    mileage: {
      number: 186400,
      unit: 'km'
    },
    fuelType: 'Petrol',
    pictureUrl: 'http://localhost:3001/car.svg'
  }
]

describe('Cars', () => {
  DuckTest(Cars)

  describe('Actions creators', () => {
    describe('setFetching', () => {
      test('Should have the correct action type', () => {
        const action = Cars.setFetching()
        expect(action.type).toEqual(Cars.SET_FETCHING)
      })

      test('Should have the correct value (false)', () => {
        const action = Cars.setFetching(false)
        expect(action.fetching).toEqual(false)
      })

      test('Should have the correct value (true)', () => {
        const action = Cars.setFetching(true)
        expect(action.fetching).toEqual(true)
      })

      test('Should set false as default', () => {
        const action = Cars.setFetching()
        expect(action.fetching).toEqual(false)
      })
    })

    describe('setPage', () => {
      test('Should have the correct action type', () => {
        const action = Cars.setPage()
        expect(action.type).toEqual(Cars.SET_PAGE)
      })

      test('Should have the correct value', () => {
        const page = 3
        const action = Cars.setPage(page)
        expect(action.page).toEqual(page)
      })

      test('Should set 1 as default', () => {
        const action = Cars.setPage()
        expect(action.page).toBe(1)
      })
    })

    describe('setCars', () => {
      test('Should have the correct action type', () => {
        const action = Cars.setCars()
        expect(action.type).toEqual(Cars.SET_CARS)
      })

      test('Should have the correct value', () => {
        const action = Cars.setCars(MOCK_CARS)
        expect(action.cars).toEqual(MOCK_CARS)
      })

      test('Should set [] as default', () => {
        const action = Cars.setCars()
        expect(action.cars).toEqual([])
      })
    })

    describe('setResultsCount', () => {
      test('Should have the correct action type', () => {
        const action = Cars.setResultsCount()
        expect(action.type).toEqual(Cars.SET_RESULTS_COUNT)
      })

      test('Should have the correct value', () => {
        const resultsCount = 6
        const action = Cars.setResultsCount(resultsCount)
        expect(action.resultsCount).toEqual(resultsCount)
      })

      test('Should set 0 as default', () => {
        const action = Cars.setResultsCount()
        expect(action.resultsCount).toEqual(0)
      })
    })
  })

  describe('Reducer', () => {
    describe('setFetching', () => {
      test('Should set the `fetching` property to the value passed', () => {
        const newState = Cars.default(
          Cars.defaultState,
          Cars.setFetching(true)
        )
        expect(newState.fetching).toBe(true)
      })
    })

    describe('setCars', () => {
      test('Should set the `cars` property to the value passed', () => {
        const newState = Cars.default(
          Cars.defaultState,
          Cars.setCars(MOCK_CARS)
        )
        expect(newState.cars).toEqual(MOCK_CARS)
      })

      test('Should set the `carsMap` property', () => {
        const newState = Cars.default(
          Cars.defaultState,
          Cars.setCars(MOCK_CARS)
        )
        var expectedMap = new Map()
        MOCK_CARS.forEach(car => {
          expectedMap = expectedMap.set(car.stockNumber, car)
        })
        expect(newState.carsMap).toEqual(expectedMap)
      })
    })

    describe('setPage', () => {
      test('Should set the `page` property to the value passed', () => {
        const page = 4
        const newState = Cars.default(
          Cars.defaultState,
          Cars.setPage(page)
        )
        expect(newState.page).toEqual(page)
      })
    })

    describe('setResultsCount', () => {
      test('Should set the `resultsCount` property to the value passed', () => {
        const resultsCount = 4
        const newState = Cars.default(
          Cars.defaultState,
          Cars.setResultsCount(resultsCount)
        )
        expect(newState.resultsCount).toEqual(resultsCount)
      })
    })
  })
})
