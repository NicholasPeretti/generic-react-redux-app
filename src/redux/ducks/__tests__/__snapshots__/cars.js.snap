// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cars defaultState should match snapshot 1`] = `
Object {
  "cars": Array [],
  "carsMap": Immutable.Map {},
  "error": null,
  "fetching": false,
  "mileageSort": null,
  "page": 1,
  "pageSize": 10,
  "resultsCount": 0,
}
`;
